test_plans/execution_model.asciidoc

```diff
@@ -0,0 +1,122 @@
+is selected on the CTS command line.
+
+== Tests
+
+=== Execution order of three command groups submitted to the same queue
+
+Create 2 buffers for large arrays. Submit 2 commands that have some long loops and then fill these buffers with expected values via accessors in loop.
+Submit command that copy content of these buffers to result buffer.
+
+[source,c++]
+----
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc1(buffer1, cgh);
+  cgh.single_task([=] {
+    // long loop
+    for (int i = SIZE; i >= 0; i--) acc1[i] = i;
+  });
+});
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc2(buffer2, cgh);
+  cgh.single_task([=] {
+    // long loop
+    for (int i = SIZE; i >= 0; i--) acc2[i] = i;
+  });
+});
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc1(buffer1, cgh);
+  sycl::accessor<int, 1> acc2(buffer2, cgh);
+  sycl::accessor<int, 1> res_acc1(res_buffer1, cgh);
+  sycl::accessor<int, 1> res_acc2(res_buffer2, cgh);
+  cgh.parallel_for(sycl::range<1>(SIZE), [=](id<1> index) {
+    res_acc1[index] = acc1[index];
+    res_acc2[index] = acc2[index];
+  });
+});
+----
+
+Check that all elements in result buffer are equal to expected values which means that requisites for third command were satisfied before it started to execute.
+
+=== Execution order of three command groups submitted to the different queues
+
+Same as previous tests but each command is run on different queue with diffrent context.
+
+=== Requirements on overlapping vs non-overlapping sub-buffer
+
+This check is skipped if device doesn't support `aspect::usm_atomic_device_allocations`.
+
+Create buffer on array, and two non-overlapping sub-buffers
+
+[source,c++]
+----
+buffer<int, 1> sub_buf1(buffer, 0, 5);
+buffer<int, 1> sub_buf2(buffer, 10, 5);
+int* pflag = sycl::malloc_device<int>(1, queue);
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc1(sub_buf1, cgh);
+  cgh.single_task([=] {
+    sycl::atomic_ref<int> rflag(*pflag);
+    // long loop
+    for (int i = SIZE; i >= 0; i--) acc1[i] = i;
+    rflag = 1;
+  });
+});
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc2(sub_buf2, cgh);
+  cgh.single_task([=] {
+    sycl::atomic_ref<int> rflag(*pflag);
+    for (int i = SIZE; i >= 0; i--) acc2[i] = i;
+    rflag = 2;
+  });
+});
+queue.submit([&](handler& cgh) {
+  sycl::accessor<int, 1> acc(buffer, cgh);
+  sycl::accessor<int, 1> res_acc(res_buffer, cgh);
+  sycl::accessor<int, 1> flag_res_acc(flag_res_buffer, cgh);
+  copy(acc1, res_acc);
+  cgh.parallel_for(SIZE, [=](auto item) {
+    res_acc[item.get_id()] = acc[item.get_id()];
+    flag_res_acc[item.get_id()] = *pflag;
+  });
+});
+----
+
+Check that *pflag was equal to 1 for all work items, that means first two commands executes independently, and check that all elements in result buffer are equal to expected values which means that requisites for third command were satisfied before it started to execute.
```

Comment:
I don't understand this test.  It seems like `*pflag` could be either `1` or `2` because the order of the first two kernels is indeterminate.  (In fact, they could even run in parallel.)

What is this test trying to validate?

Reply:
The idea was that since they executes independently the one without long loop is expected to finish first, but now I think we still can't guarantee that. Do you think there is other way to check that kernels with assess to different parts of buffer execute independently or we should discard this check completely?

Reply:
The spec does not require the first two kernels to run concurrently, it only says that it is legal to run them concurrently if the implementation chooses.  Therefore, I think it's hard to write a test for that.

However, we could write a related test.  The spec says that kernels with overlapping sub-buffers can **not** be run concurrently.  We could create a test that has overlapping sub-buffers, and then verify that the kernels run in a specific order.

Reply:
Changed test case to check order of kernels with overlapping sub-buffers in https://github.com/KhronosGroup/SYCL-CTS/pull/632/commits/04f949f0231db3d479e7fe828581afcb33ca08b3